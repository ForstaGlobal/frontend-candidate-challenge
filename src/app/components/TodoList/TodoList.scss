@import "../../styles/colors.scss";
.todolist-container {
    background-color: white;
    margin-top: 1rem;
    border-radius: 0.25rem; 
    height: 60vh;
    display: flex;
    flex-direction: column;
    justify-content: space-between;
}

.todos-container {
    overflow: auto;
    height: 100%;
    padding: 0;
}

.todo-scroll { 
    &::-webkit-scrollbar {
        width: 5px;
    }
 
    &::-webkit-scrollbar-track {
        background: #f1f1f1;
    }
 
    &::-webkit-scrollbar-thumb {
        background: #888;
    }
 
    &::-webkit-scrollbar-thumb:hover {
        background: #555;
        background: $light-pink;
    }
}

