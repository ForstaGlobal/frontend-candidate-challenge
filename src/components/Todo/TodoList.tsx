import { Todo } from "../../interfaces/types";
import { CiEdit } from "react-icons/ci";
import { CiTrash } from "react-icons/ci";
import { FcTodoList } from "react-icons/fc";
import Task from "./Task";

type TodoListProps = {
  todos: Todo[];
  onDeleteTask: (id: string) => void;
  onEditTask: (id: string) => void;
  onToggleComplete: (id: string, isDone: boolean) => void;
};
export const TodoList = ({
  todos,
  onDeleteTask,
  onEditTask,
  onToggleComplete,
}: TodoListProps) => {
  const EmptyListState = () => (
    <div className="no-tasks">
      <FcTodoList size={50} />
      <p>You have no tasks. Add one now!</p>
    </div>
  );

  const deleteTask = (todo: Todo) => {
    const res = window.confirm(`Are you sure you want to delete ${todo.task}`);
    if (!res) {
      return;
    }
    onDeleteTask(todo.id);
  };

  return (
    <ul className="todo-list-container">
      {todos.length ? (
        todos.map((item, i) => (
          <li key={i}>
            <Task
              todo={item}
              onToggleComplete={(isComplete) =>
                onToggleComplete(item.id, isComplete)
              }
            />
            <div className="todo-list-item-actions">
              {!item.isDone && (
                <span
                  className="todo-list-item-edit"
                  onClick={() => onEditTask(item.id)}
                >
                  <CiEdit size={24} />
                </span>
              )}
              <span
                className="todo-list-item-delete"
                onClick={() => deleteTask(item)}
              >
                <CiTrash size={24} />
              </span>
            </div>
          </li>
        ))
      ) : (
        <EmptyListState />
      )}
    </ul>
  );
};
