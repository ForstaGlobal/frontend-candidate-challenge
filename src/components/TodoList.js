import React from "react";

export const TodoList = (props) => {
    function toggleDone(event, item) {
        props.onToggle(item);
    }

    function deleteIt(event, item) {
        props.onDelete(item);
    }

    return (
    <ul className="todoList">
      {props.todos.map((item) => (
        <li key={item.id}>
          <span
            data-testid={`todo${item.id}`}
            style={{ textDecoration: item.done ? 'line-through': 'none'}}
            onClick={ (event) => toggleDone(event, item) }
          >
            {item.text}
          </span>
          <span
            className="delBtn"
            onClick={ (event) => deleteIt(event, item) }
          >
            x
          </span>
        </li>
      ))}
    </ul>
  );
};
