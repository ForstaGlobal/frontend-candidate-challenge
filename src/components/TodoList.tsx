import React from 'react';
import { Todo } from '../types';

type TodoListProps = {
  todos: any[];
};
export const TodoList = ({ todos }: TodoListProps) => {
  return (
    <ul className='todoList'>
      {todos.map((todo: Todo, i) => (
        <li key={todo.id}>
          <span data-testid={`todo${todo.id}`}>{todo.text}</span>
        </li>
      ))}
    </ul>
  );
};
