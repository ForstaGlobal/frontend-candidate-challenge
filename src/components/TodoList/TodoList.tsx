import React from "react";
import { Todo } from "../../types";
import TodoItem from "../TodoItem";

type TodoListProps = {
  todos: Todo[];
  setTodos: React.Dispatch<React.SetStateAction<Todo[]>>;
};
export const TodoList = ({ todos, setTodos }: TodoListProps) => {
  return (
    <ul className="todoList">
      {todos.map((todo) => (
        <TodoItem todo={todo} key={todo.id} />
      ))}
    </ul>
  );
};
