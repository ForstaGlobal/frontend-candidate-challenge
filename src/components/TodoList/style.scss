.list-container {
  width: 100%;
  display: flex;
  flex-direction: column;
  gap: 10px;

  &__count {
    display: flex;
    justify-content: space-between;
  }
}
