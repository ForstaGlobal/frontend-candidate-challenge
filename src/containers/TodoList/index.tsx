import React from "react";
import AddTask from "./components/AddTask/AddTask"
import ToDoList from "./components/ToDoList";
import { useAppSelector, useAppDispatch } from '../../app/hooks';
import {
    addTask,
    updateTask,
    selectTodos,
} from './store/index';

import "./ToDoList.scss"

export const TodoList = () => {
    const todos = useAppSelector(selectTodos);
    const dispatch = useAppDispatch();
    const doneTasks = todos.filter((task: any) => { if (task.done) return task; return 0 })
    const notDoneTasks = todos.filter((task: any) => { if (!task.done) return task; return 0 })

    const handleAddTask = (nameTask: string) => {
        const newTask = [...todos, {
            id: todos.length + 1,
            name: nameTask,
            done: false,
            editable: false
        }]
        dispatch(addTask(newTask))
    }

    const handleDeleteTask = (id: number) => {
        const deleteTask = todos.filter((task: any) => {
            if (task.id !== id) {
                return task;
            }
            return 0;
        })
        dispatch(updateTask(deleteTask))
    }

    const handleUpdateTask = (e: { target: { checked: React.SetStateAction<any>; }; }, id: number) => {
        const editTask = todos.map((task: any) => {
            if (task.id === id) {
                return { ...task, done: e.target.checked }
            }
            return task;
        })
        dispatch(updateTask(editTask))
    }

    const handleEdit = (id: number) => {
        const editTask = todos.map((task: any) => {
            if (task.id === id) {
                return { ...task, editable: !task.editable }
            }
            return task;
        })
        dispatch(updateTask(editTask))
    }

    const handleChangeUpdate = (id: number, newName: string) => {
        const newTask = todos.map((task: any) => {
            if (task.id === id) {
                return { ...task, name: newName, editable: !task.editable }
            }
            return task
        })
        dispatch(updateTask(newTask))
    }

    return (
        <div>
            <AddTask onAdd={handleAddTask} todos={todos} />
            <ToDoList title="Not Done" todos={notDoneTasks} onUpdate={handleUpdateTask} onChange={handleChangeUpdate} onEdit={handleEdit} onDelete={handleDeleteTask} />
            <ToDoList title="Done" todos={doneTasks} onUpdate={handleUpdateTask} onChange={handleChangeUpdate} onEdit={handleEdit} onDelete={handleDeleteTask} />
        </div>
    );
};
