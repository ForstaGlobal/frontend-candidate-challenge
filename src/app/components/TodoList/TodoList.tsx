import { useSelector } from "react-redux";
import { RootState } from "../../redux/store";
import "./TodoList.scss";
import TodoItem from "../TodoItem/TodoItem";
import { useMemo } from "react";

function TodoList() {
  const todos = useSelector((state: RootState) => state.todo.todos);
  const searchQuery = useSelector((state: RootState) => state.todo.searchQuery);
  const filterCategory = useSelector(
    (state: RootState) => state.todo.filterCategory
  );

  const filteredTodos = useMemo(() => {
    return todos.filter(
      (todo) =>
        todo.category.toLowerCase().includes(filterCategory.toLowerCase()) &&
        (todo.task.toLowerCase().includes(searchQuery.toLowerCase()) ||
          todo.description?.toLowerCase().includes(searchQuery.toLowerCase()) ||
          todo.category.toLowerCase().includes(searchQuery.toLowerCase()))
    );
  }, [todos, filterCategory, searchQuery]);

  return (
    <div className="todolist-container">
      {filteredTodos.length === 0 ? (
        <p>No todo items found</p>
      ) : (
        <ul className="todos-container">
          {filteredTodos.map((todo) => (
            <TodoItem todo={todo} key={todo.id} />
          ))}
        </ul>
      )}
    </div>
  );
}

export default TodoList;
