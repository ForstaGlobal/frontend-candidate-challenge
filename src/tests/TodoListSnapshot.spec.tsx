/** @jest-environment jsdom */
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import TodoList from '../components/TodoList';

describe('TodoList Component', () => {

  it('should render correctly and match snapshot', () => {
    const wrapper = shallow(<TodoList />);
    expect(toJson(wrapper)).toMatchSnapshot();
  });
});
