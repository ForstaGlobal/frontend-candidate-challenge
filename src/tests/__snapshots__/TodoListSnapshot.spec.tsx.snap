// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TodoList Component should render correctly and match snapshot 1`] = `
<div>
  <ToggleButtons
    setFilter={[Function]}
  />
  <ForwardRef(List) />
</div>
`;
